cxx/hirm.cc

```diff
@@ -332,37 +331,67 @@ void HIRM::sample_and_incorporate_relation(std::mt19937* prng,
         },
         get_relation(r));
   }
-  std::visit([&](auto rel) { rel->incorporate_sample(prng, items); },
+  std::ostringstream ss;
+  std::visit([&](auto rel) { ss << rel->incorporate_sample(prng, items); },
              get_relation(r));
+  return ss.str();
 }
 
-void HIRM::sample_and_incorporate(std::mt19937* prng, int n) {
+T_encoded_observations HIRM::sample_and_incorporate(std::mt19937* prng, int n) {
+  T_encoded_observations obs;
   std::map<std::string, CRP> domain_crps;
+  // Initialize the domain_crps from all the data across all the IRM's.
+  // If sample_and_incorporate ends up being used a lot more, we should
+  // move the domain_crps into the HIRM and maintain them when incorporating
+  // rather than recalculating them here.
+  for (const auto& [unused_cluster_id, irm] : irms) {
+    for (const auto& [domain_name, domain] : irm->domains) {
+      for (const auto& [item, table] : domain->crp.assignments) {
+        domain_crps[domain_name].incorporate(item, table);
```

Comment:
Recall that in `domain->crp.assignments`, `item` is the entity ID and `table` is the cluster to which the entity belongs. In `domain_crps`, the "tables" are the entity IDs and the customers/"items" are just values from a counter of observations of that entity. So this should be:

```
int crp_item = domain_crps[domain_name].assignments.size();
domain_crps[domain_name].incorporate(crp_item, item);
```
similar to lines 370-371.

Reply:
That fixed it, thanks.